distributed/shuffle/_worker_extension.py

```diff
@@ -193,6 +155,359 @@ def heartbeat(self) -> dict[str, Any]:
+        self.new = new
+        partitions_of = defaultdict(list)
+        for part, addr in worker_for.items():
+            partitions_of[addr].append(part)
+        self.partitions_of = dict(partitions_of)
+        self.worker_for = worker_for
+        self._mapping = self._map_old_to_new(old, new)
+        self._old_to_new = _old_to_new(old, new)
+
+    def _map_old_to_new(
+        self, old: ChunkedAxes, new: ChunkedAxes
+    ) -> dict[NIndex, list[tuple[NIndex, NIndex, NSlice]]]:
+
+        # intersections is contains the new individual chunks as combined slices
+        # of the old chunks.
+        #
+        # Each chunk consists of all the n-dimensional slices of the old chunks
+        # that make up the new chunk.
+        #
+        # Each n-dimensional slice contains n tuples consisting of the index of
+        # the old chunk on the n-th dimension and the slice along that dimension.
+        intersections = intersect_chunks(old, new)
+
+        new_indices = product(*(range(len(c)) for c in new))
+
+        ndim = len(old)
+
+        old_to_new = defaultdict(list)
+
+        for new_index, new_chunk in zip(new_indices, intersections):
+            # sub-dimensions of new chunk composed of slices
+            subdims = [
+                len({slice[dim][0] for slice in new_chunk}) for dim in range(ndim)
+            ]
+
+            subdim_indices = product(*(range(dim) for dim in subdims))
+
+            for subdim_index, nchunkslice in zip(subdim_indices, new_chunk):
+                old_index, slices = zip(*nchunkslice)
+                old_to_new[old_index].append((new_index, subdim_index, slices))
+        return old_to_new
+
+    async def _receive(self, data: list[tuple[tuple[NIndex, NIndex], bytes]]) -> None:
+        self.raise_if_closed()
+
+        filtered = []
+        for d in data:
+            if d[0] not in self.received:
+                filtered.append(d)
+                self.received.add(d[0])
+                self.total_recvd += sizeof(d)
+        del data
+        if not filtered:
+            return
+        try:
+            buffers = await self.offload(self._reserialize_buffers, filtered)
+            del filtered
+            await self._write_to_disk(buffers)
+        except Exception as e:
+            self._exception = e
+            raise
+
+    def _reserialize_buffers(
+        self, data: list[tuple[tuple[NIndex, NIndex], bytes]]
+    ) -> dict[NIndex, list[bytes]]:
+        result = defaultdict(list)
+        for d in data:
+            result[d[0][0]].append(d[1])
+        return result
+
+    async def add_partition(self, data: np.ndarray, input_partition: NIndex) -> int:
+        self.raise_if_closed()
+        if self.transferred:
+            raise RuntimeError(f"Cannot add more partitions to shuffle {self}")
+
+        def _() -> dict[str, list[tuple[tuple[NIndex, NIndex], bytes]]]:
+
+            out: dict[str, list[tuple[tuple[NIndex, NIndex], bytes]]] = defaultdict(
```

Comment:
I would appreciate a couple of words somewhere describing the indexing scheme

Reply:
Done